### PR TITLE
add discrete_range distributions, svd_U and svd_V

### LINKED ISSUE #750: Expose the discrete_range functions
The following signatures (introduced in https://github.com/stan-dev/math/pull/1680 and https://github.com/stan-dev/math/pull/1716) should be exposed by the compiler:

real discrete_range_log(int, int, int)
real discrete_range_log(int[], int[], int[])
real discrete_range_lpmf(int, int, int)
real discrete_range_lpmf(int[], int[], int[])
real discrete_range_cdf(int, int, int)
real discrete_range_cdf(int[], int[], int[])
real discrete_range_cdf_log(int, int, int)
real discrete_range_cdf_log(int[], int[], int[])
real discrete_range_lcdf(int, int, int)
real discrete_range_lcdf(int[], int[], int[])
real discrete_range_lccdf(int, int, int)
real discrete_range_lccdf(int[], int[], int[])
real discrete_range_ccdf_log(int, int, int)
real discrete_range_ccdf_log(int[], int[], int[])
int discrete_range_rng(int, int)
int[] discrete_range_rng(int[], int[])

### PR DESCRIPTION
**Pull request by rok-cesnovar**
Fixes #750 

by adding support for the following signatures:
```
real discrete_range_ccdf_log(int, int, int)
real discrete_range_ccdf_log(int, int, int[])
real discrete_range_ccdf_log(int, int[], int)
real discrete_range_ccdf_log(int, int[], int[])
real discrete_range_ccdf_log(int[], int, int)
real discrete_range_ccdf_log(int[], int, int[])
real discrete_range_ccdf_log(int[], int[], int)
real discrete_range_ccdf_log(int[], int[], int[])
real discrete_range_cdf(int, int, int)
real discrete_range_cdf(int, int, int[])
real discrete_range_cdf(int, int[], int)
real discrete_range_cdf(int, int[], int[])
real discrete_range_cdf(int[], int, int)
real discrete_range_cdf(int[], int, int[])
real discrete_range_cdf(int[], int[], int)
real discrete_range_cdf(int[], int[], int[])
real discrete_range_cdf_log(int, int, int)
real discrete_range_cdf_log(int, int, int[])
real discrete_range_cdf_log(int, int[], int)
real discrete_range_cdf_log(int, int[], int[])
real discrete_range_cdf_log(int[], int, int)
real discrete_range_cdf_log(int[], int, int[])
real discrete_range_cdf_log(int[], int[], int)
real discrete_range_cdf_log(int[], int[], int[])
real discrete_range_lccdf(int, int, int)
real discrete_range_lccdf(int, int, int[])
real discrete_range_lccdf(int, int[], int)
real discrete_range_lccdf(int, int[], int[])
real discrete_range_lccdf(int[], int, int)
real discrete_range_lccdf(int[], int, int[])
real discrete_range_lccdf(int[], int[], int)
real discrete_range_lccdf(int[], int[], int[])
real discrete_range_lcdf(int, int, int)
real discrete_range_lcdf(int, int, int[])
real discrete_range_lcdf(int, int[], int)
real discrete_range_lcdf(int, int[], int[])
real discrete_range_lcdf(int[], int, int)
real discrete_range_lcdf(int[], int, int[])
real discrete_range_lcdf(int[], int[], int)
real discrete_range_lcdf(int[], int[], int[])
real discrete_range_log(int, int, int)
real discrete_range_log(int, int, int[])
real discrete_range_log(int, int[], int)
real discrete_range_log(int, int[], int[])
real discrete_range_log(int[], int, int)
real discrete_range_log(int[], int, int[])
real discrete_range_log(int[], int[], int)
real discrete_range_log(int[], int[], int[])
real discrete_range_lpmf(int, int, int)
real discrete_range_lpmf(int, int, int[])
real discrete_range_lpmf(int, int[], int)
real discrete_range_lpmf(int, int[], int[])
real discrete_range_lpmf(int[], int, int)
real discrete_range_lpmf(int[], int, int[])
real discrete_range_lpmf(int[], int[], int)
real discrete_range_lpmf(int[], int[], int[])
int discrete_range_rng(int, int)
int[] discrete_range_rng(int, int[])
int[] discrete_range_rng(int[], int)
int[] discrete_range_rng(int[], int[])
```
and
```
matrix svd_U(matrix);
matrix svd_V(matrix);
```

## Release notes

Added the discrete_range distribution.

## Copyright and Licensing

By submitting this pull request, the copyright holder is agreeing to 
license the submitted work under the BSD 3-clause license (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

**Comment by rok-cesnovar** [comment]
Anyone up for reviewing this?

**Comment by rok-cesnovar** [comment]
This now also fixes #790 

Ready for a look.

File: test/integration/good/function-signatures/distributions/univariate/discrete/discrete_range/pretty.expected [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+  transformed_param_real = discrete_range_lccdf(d_int_array| d_int, d_int_array);
+  transformed_param_real = discrete_range_lccdf(d_int_array| d_int_array, d_int);
+  transformed_param_real = discrete_range_lccdf(d_int_array| d_int_array, d_int_array);
+}
+model {
+  y_p ~ normal(0, 1);
+}
+
+
+Semantic error in 'discrete_range_lccdf.stan', line 7, column 26 to column 67:
```
Comment (bbbales2) [inline test/integration/good/function-signatures/distributions/univariate/discrete/discrete_range/pretty.expected:79]: Is this an actual error?

**Review by bbbales2 — CHANGES_REQUESTED** [review]
I see `_log` signatures in stan_math_sigs.expected but they're not tested to compile. Dunno if this is expected.

I see some errors in one of the pretty prints too.

Otherwise I think everything looks good.

I see tests that all the lpmf/lcdf/cdf/lccdf signatures for discrete_range compile in transformed data and transformed parameters. Similarly I see that `svd_U` and `svd_V` are tested in the same places.

For the rngs it looks like all signatures are tested in generated quantities, which makes sense.

File: test/integration/good/function-signatures/distributions/univariate/discrete/discrete_range/pretty.expected (hunk 1 above)
Comment (rok-cesnovar) [reply test/integration/good/function-signatures/distributions/univariate/discrete/discrete_range/pretty.expected:79]: Yes it is. Let me check.

**Comment by rok-cesnovar** [comment]
I see _log signatures in stan_math_sigs.expected but they're not tested to compile. Dunno if this is expected.

When you add a distribution in stanc3 it automatically also generated those sigs. Do we not have _log support in Math for discrete_range?

File: test/integration/good/function-signatures/distributions/univariate/discrete/discrete_range/pretty.expected (hunk 1 above)
Comment (rok-cesnovar) [reply test/integration/good/function-signatures/distributions/univariate/discrete/discrete_range/pretty.expected]: This was legit, I did not use vertical bars for lccdf calls in trans. data.

**Review by bbbales2 — APPROVED** [review]

**Pull request merged**